include/vast/Conversion/ABI/AggregateTypes.hpp

```diff
@@ -0,0 +1,286 @@
+// Copyright (c) 2023-present, Trail of Bits, Inc.
+
+#pragma once
+
+#include "vast/Util/Warnings.hpp"
+
+VAST_RELAX_WARNINGS
+VAST_UNRELAX_WARNINGS
+
+namespace vast::conv::abi
+{
+    /* Handles aggregate type reconstruction. */
+
+    // TODO(conv:abi): Parametrize by materializer, which will resolve what dialect
+    //                is the target
+    //                 * hl
+    //                 * ll-vars
+    //                 * llvm
+    template< typename pattern, typename op_t >
+    struct field_allocator
+    {
+        using self_t = field_allocator< pattern, op_t >;
+
+        struct state_t
+        {
+            std::size_t arg_idx = 0;
+            std::size_t offset = 0;
+
+            const pattern &parent;
+            op_t abi_op;
+
+            state_t(const pattern &parent, op_t abi_op)
+                : parent(parent), abi_op(abi_op)
+            {}
+
+            state_t( const state_t &) = delete;
+            state_t( state_t &&) = delete;
+
+            state_t &operator=(state_t) = delete;
+
+            auto arg() { return abi_op.getOperands()[arg_idx]; };
+            auto bw(auto w) { return parent.bw(w); };
+
+            bool fits(mlir_type type)
+            {
+                return bw(arg()) >= offset + bw(type);
+            };
+
+            void advance()
+            {
+                ++arg_idx;
+                offset = 0;
+            }
+
+            // TODO(conv:abi): Possibly add some debug prints to help us
+            //                 debug some weird corner cases in the future?
+            mlir::Value allocate(mlir_type type, auto &rewriter)
+            {
+                auto start = offset;
+                offset += bw(type);
+                if (bw(type) == bw(arg()))
+                    return arg();
+
+                return rewriter.template create< ll::Extract >(
+                        abi_op.getLoc(), type, arg(),
+                        start, offset);
+            };
+        };
+
+        state_t &state;
+        vast_module mod;
+        std::vector< mlir::Value > partials;
+
+        field_allocator(state_t &state, vast_module mod)
+            : state(state),
+              mod(mod)
+        {}
+
+        static state_t mk_state(const pattern &parent, op_t abi_op)
+        {
+            return state_t(parent, abi_op);
+        }
+
+        bool needs_nesting(mlir_type type) const
```

Comment:
Isn't this in fact obsolete as of `mlir::AttrTypeRewriter`?

Reply:
Well this is sadly not that simple, since I am pretty sure arrays will be treated same way as any scalar value and we cannot nest here. I will update the docs and merge this with the same functionality below to have it only at one place, but I do not think we can easily now re-use some existing MLIR abstraction.